File: CS/InputManager.cs
/*
 * Team 3
 * AJ Downey, Joshua Barbee, Wei Wei Chien
 */

using System.Security;
using System.Text.RegularExpressions;

namespace Defend483CS
{

    internal class InputManager
    {

        internal const int _maxStringLength = 50;

        private const string _logFilename =
            "log.txt";
        private const string _passwordFilename =
            "pass.byte";

        private const string _backupProgramFolder =
            @"tcss360\defend";
        private const string _backupLogPath =
            _backupProgramFolder + @"\" + _logFilename;
        private const string _backupPasswordPath =
            _backupProgramFolder + @"\" + _passwordFilename;

        private static readonly string _programFolder =
            Path.Combine(
                Environment.GetFolderPath(
                    Environment.SpecialFolder.ApplicationData
                ),
                _backupProgramFolder
            );
        private static readonly string _logPath =
            Path.Combine(_programFolder, _logFilename);
        private static readonly string _passwordPath =
            Path.Combine(_programFolder, _passwordFilename);

        private static readonly string _invalidInput =
            "Invalid input. Try again." + Environment.NewLine;
        private const string _redirectionAttempt =
            "File redirection is not allowed.";
        private const string _programFolderError =
            "The application cannot acquire the file system resources to run.";
        private const string _inputTooLong =
            "The provided input was too long.";


        internal static bool EnsureNoRedirection()
        {
            if (Console.IsOutputRedirected)
            {
                // Don't attempt to print to console, since may be redirected
                // to unreliable output that may cause other exception
                LogError(_redirectionAttempt);
                return false;
            }
            if (Console.IsInputRedirected)
            {
                Console.WriteLine(_redirectionAttempt);
                LogError(_redirectionAttempt);

                return false;
            }

            return true;
        }

        internal static (T, T) ReadTwo<T>(in string theTitle,
                                          in string theFirstSubtitle,
                                          in string theSecondSubtitle,
                                          in Func<string, T> theReadFunction)
        {
            Console.WriteLine(theTitle);

            T first = theReadFunction(theFirstSubtitle);
            T second = theReadFunction(theSecondSubtitle);

            return (first, second);
        }

        internal static string ReadStringUntilValid(in string thePrompt,
                                                    in string theRegex = "^.*$",
                                                    in string? theTitle = null,
                                                    in Predicate<string>? theValidationFunction = null)
        {
            string? result;
            Predicate<string> validationFunction = theValidationFunction ??
                                                   ((str) => true);

            if (!string.IsNullOrEmpty(theTitle))
            {
                Console.WriteLine(theTitle);
            }

            while (true)
            {
                Console.WriteLine(thePrompt);

                result = ReadLineAndPrintBlankLine();
                if (result != null &&
                    Regex.IsMatch(result, theRegex) &&
                    validationFunction(result))
                {
                    return result;
                }

                Console.WriteLine(_invalidInput);
            }
        }

        internal static string GetPasswordPath()
        {
            return MakeProgramDirectory() ? _passwordPath : _backupPasswordPath;
        }

        internal static void LogError(in Exception theException)
        {
            LogError(theException.ToString());
        }

        internal static void CloseUnexpectedly(in Exception theException,
                                               in ExitCode theExitCode)
        {
            LogError(theException);
            CloseUnexpectedly(theExitCode);
        }

        private static string? ReadLineAndPrintBlankLine()
        {
            string? line = null;

            try
            {
                line = Console.ReadLine(); // May return null w/o exception
                Console.WriteLine();
            }
            catch (Exception e) when (e is OutOfMemoryException or
                                      ArgumentOutOfRangeException)
            {
                Console.WriteLine(); // Exception before first cw, so again
                Console.WriteLine(_inputTooLong);
            }
            catch (IOException e)
            {
                // No redirection, so should only temporarily occur unless
                // serious problem in environment (so expect to not close
                // program)
                LogError(e);
                // Slight formatting discrepancy with no cw, but
                // cw excluded to avoid printing large number of
                // lines if repeated IOExceptions without ever
                // prompting user for input
            }

            return line;
        }

        private static string GetLogPath()
        {
            return MakeProgramDirectory() ? _logPath : _backupLogPath;
        }

        private static bool MakeProgramDirectory()
        {
            if (MakeProgramDirectory(_programFolder))
            {
                return true;
            }
            if (MakeProgramDirectory(_backupProgramFolder))
            {
                return false;
            }

            // Neither folder can be used, so nowhere to preserve error
            // information after program exits
            ProgramFolderFailure();
            return false; // Unreachable
        }

        private static bool MakeProgramDirectory(in string theDirectory)
        {
            try
            {
                if (Directory.CreateDirectory(theDirectory).Exists)
                {
                    return true;
                }
            }
            catch (Exception e) when (e is IOException or
                                      UnauthorizedAccessException or
                                      ArgumentException or
                                      NotSupportedException)
            {// Indicate failure by returning false. To avoid recursive loop
            }// when errors cannot be logged, don't attempt to log error

            // Return false if directory doesn't exist for some reason
            // according to DirectoryInfo.Exists or if exception thrown by
            // CreateDirectory otherwise indicates specified directory not
            // available
            return false;
        }

        private static void LogError(in string theError)
        {
            try
            {
                File.AppendAllText(
                    GetLogPath(),
                    Environment.NewLine +
                    DateTime.Now.ToString("F") + ':' + Environment.NewLine +
                    theError + Environment.NewLine
                );
            }
            catch (Exception e) when (e is IOException or
                                      UnauthorizedAccessException or
                                      ArgumentException or
                                      NotSupportedException or
                                      SecurityException)
            {
                // Preliminary check for at least one program folder option
                // working avoids this in as many situations as possible (short
                // of simply continuing to let the program run when errors
                // occur and cannot be logged)
                ProgramFolderFailure();
            }
        }

        private static void ProgramFolderFailure()
        {
            // If user can only interact with program, should never be
            // encountered

            // Brief, not very revealing message written to console instead of
            // error log since error log unavailable. Check for redirection
            // needed since may be called from EnsureNoRedirection
            if (!Console.IsOutputRedirected)
            {
                Console.WriteLine(_programFolderError);
            }

            CloseUnexpectedly(ExitCode.FileSystemError);
        }

        private static void CloseUnexpectedly(in ExitCode theExitCode)
        {
            Environment.Exit((int) theExitCode);
        }
    }
}

File: Java/defend/InputManager.java
/*
 * Team 3
 * AJ Downey, Joshua Barbee, Wei Wei Chien
 */

package defend;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Date;
import java.util.NoSuchElementException;
import java.util.Scanner;
import java.util.function.Predicate;

public class InputManager {

    static final int MAX_STRING_LENGTH = 50;

    private static final String LOG_FILENAME = "log.txt";
    private static final String PASSWORD_FILENAME = "pass.byte";

    private static final String BACKUP_PROGRAM_FOLDER =
            "tcss360\\defend";
    private static final String BACKUP_LOG_PATH =
            BACKUP_PROGRAM_FOLDER + '\\' + LOG_FILENAME;
    private static final String BACKUP_PASSWORD_PATH =
            BACKUP_PROGRAM_FOLDER + '\\' + PASSWORD_FILENAME;

    private static final String PROGRAM_FOLDER =
            System.getenv("APPDATA") + '\\' + BACKUP_PROGRAM_FOLDER;
    private static final String LOG_PATH =
            PROGRAM_FOLDER + '\\' + LOG_FILENAME;
    private static final String PASSWORD_PATH =
            PROGRAM_FOLDER + '\\' + PASSWORD_FILENAME;

    private static final String INVALID_INPUT =
            "Invalid input. Try again.\n";
    private static final String REDIRECTION_ATTEMPT =
            "File redirection is not allowed.";
    private static final String PROGRAM_FOLDER_ERROR =
            "The application cannot acquire the file system resources to run.";
    private static final String UNEXPECTED_CLOSE =
            "The application was forced to close unexpectedly.";

    private static Scanner INPUT = new Scanner(System.in);


    static boolean ensureNoRedirection() {
        // Require input and output to be connected to a terminal
        if (System.console() == null) {
            System.out.println(REDIRECTION_ATTEMPT);
            logError(REDIRECTION_ATTEMPT);

            return false;
        }

        return true;
    }

    static void printInputInvalid() {
        System.out.println(INVALID_INPUT);
    }

    static String readLineAndPrintBlankLine() {
        String line;
        try {
            line = INPUT.nextLine();
        } catch (NoSuchElementException e) {
            INPUT = new Scanner(System.in);
            line = null;
        }

        System.out.println();
        return line;
    }

    static String readStringWithRegex(final String theTitle,
                                      final String thePrompt,
                                      final String theRegex,
                                      final Predicate<String> theValidationFunction) {
        return readStringWithRegex(
                theTitle, thePrompt, theRegex, theValidationFunction, true
        );
    }

    static String readStringWithRegex(final String thePrompt,
                                      final String theRegex,
                                      final Predicate<String> theValidationFunction) {
        return readStringWithRegex(
                "", thePrompt, theRegex, theValidationFunction, false
        );
    }

    static String readStringWithRegex(final String theTitle,
                                      final String thePrompt,
                                      final String theRegex) {
        return readStringWithRegex(
                theTitle, thePrompt, theRegex, (str) -> true, true
        );
    }

    static String readStringWithRegex(final String theTitle,
                                      final String thePrompt,
                                      final String theRegex,
                                      final Predicate<String> theValidationFunction,
                                      final boolean theUsesTitle) {
        String result;

        if (theUsesTitle) {
            System.out.println(theTitle);
        }

        while (true) {
            System.out.println(thePrompt);

            result = readLineAndPrintBlankLine();
            if (result != null &&
                result.matches(theRegex) &&
                theValidationFunction.test(result)) {
                return result;
            }

            System.out.println(INVALID_INPUT);
        }
    }

    static String getPasswordPath() {
        return makeProgramDirectory() ? PASSWORD_PATH : BACKUP_PASSWORD_PATH;
    }

    static void logError(final Throwable theThrowable) {
        logError(theThrowable.toString());
    }

    static void unexpectedClose(final Throwable theThrowable) {
        logError(theThrowable);
        unexpectedClose();
    }

    private static String getLogPath() {
        return makeProgramDirectory() ? LOG_PATH : BACKUP_LOG_PATH;
    }

    private static boolean makeProgramDirectory() {
        final File dir = new File(PROGRAM_FOLDER);

        dir.mkdirs();
        if (!dir.exists()) {
            final File backupDir = new File(BACKUP_PROGRAM_FOLDER);

            backupDir.mkdirs();
            if (!backupDir.exists()) {
                programFolderFailure();
            }
            return false;
        }
        return true;
    }

    private static void logError(final String theError) {
        try (FileWriter fw = new FileWriter(getLogPath(), true);
             BufferedWriter bw = new BufferedWriter(fw);
             PrintWriter pw = new PrintWriter(bw)) {
            pw.println();

            pw.print(new Date());
            pw.println(':');

            pw.println(theError);
        } catch (IOException e) {
            programFolderFailure();
        }
    }

    private static void programFolderFailure() {
        // Since assuming user can only interact with system through this
        // program, should never be encountered
        if (System.console() != null) {
            System.out.println(PROGRAM_FOLDER_ERROR);
        }

        unexpectedClose();
    }

    private static void unexpectedClose() {
        if (System.console() != null) {
            System.out.println(UNEXPECTED_CLOSE);
        }

        System.exit(-1);
    }
}
